tests/lib/ocpp/v201/test_charge_point.cpp

```diff
@@ -0,0 +1,321 @@
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfResetIsAllowedCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.is_reset_allowed_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfResetCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.reset_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfStopTransactionCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.stop_transaction_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfPauseChargingCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.pause_charging_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfConnectorEffectiveOperativeStatusChangedCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.connector_effective_operative_status_changed_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfGetLogRequestCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.get_log_request_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfUnlockConnectorCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.unlock_connector_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfRemoteStartTransactionCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.remote_start_transaction_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfIsReservationForTokenCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.is_reservation_for_token_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfUpdateFirmwareRequestCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.update_firmware_request_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfSecurityEventCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.security_event_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
+}
+
+TEST_F(ChargePointFixture, K01FR02_CallbacksValidityChecksIfSetChargingProfilesCallbackExists) {
+    configure_callbacks_with_mocks();
+    callbacks.set_charging_profiles_callback = nullptr;
+
+    EXPECT_FALSE(callbacks.all_callbacks_valid());
```

Comment:
is the callback validity actually a test case for `K01FR02`? Why? It doesn't seem to be related to the requirements in the spec.

Reply:
@couryrr-afs: I agree with @shankari that this could use some context. Could you move the `all_callbacks_valid` tests that involve `set_charging_profiles_callback` to the top of these tests (if they're not there already) and include the following comment:

```c++
/*
 * K01.FR.02 states
 *
 *     "The CSMS MAY send a new charging profile for the EVSE that SHALL be used
 *      as a limit schedule for the EV."
 *
 * When using libocpp, a charging station is notified of a new charging profile
 * by means of the set_charging_profiles_callback. In order to ensure that a new
 * profile can be immediately "used as a limit schedule for the EV", a
 * valid set_charging_profiles_callback must be provided.
 *
 * As part of testing that K01.FR.02 is met, we provide the following tests that
 * confirm an OCPP 2.0.1 ChargePoint with smart charging enabled will only
 * consider its collection of callbacks valid if set_charging_profiles_callback
 * is provided.
 */
```

Following those tests, we can add a comment like the following:
```c++
/*
 * For completeness, we also test that all other callbacks are checked by
 * validate_all_profiles.
 */
```

The remaining `validate_all_profiles` tests could immediately follow this comment but with references to K01.FR.02 removed from their names.

Reply:
I could see moving the other tests to a different location in the future but think it's fine to have them here for now. If other validity tests are relevant for functional requirements in the other functional blocks (as is highly likely), they can be moved around as needed. For now, though, it's nice to keep them grouped together.